File: server/release/src/model/template.py
r"""
模板相关模型定义

:file: src/model/template.py
:author: WaterRun
:time: 2026-01-29
"""

from datetime import datetime

from pydantic import BaseModel, Field


class TemplateCodeCreate(BaseModel):
    r"""
    模板代码创建请求体，用于管理员创建模板
    """

    language: str = Field(description="编程语言")
    code: str = Field(description="代码内容")
    explanation: str | None = Field(
        default=None, description="代码说明，Markdown格式")


class TemplateCodeResponse(BaseModel):
    r"""
    模板代码响应
    """

    language: str = Field(description="编程语言")
    code: str = Field(description="代码内容")
    explanation: str | None = Field(
        default=None, description="代码说明，Markdown格式")


class TemplateCreate(BaseModel):
    r"""
    模板创建请求体，用于管理员创建模板
    """

    title: str = Field(max_length=256, description="模板标题")
    category: str = Field(max_length=64, description="模板分类")
    description: str = Field(description="模板描述，Markdown格式")
    codes: list[TemplateCodeCreate] = Field(description="多语言代码实现列表")


class TemplateUpdate(BaseModel):
    r"""
    模板更新请求体，用于管理员更新模板
    """

    title: str | None = Field(default=None, max_length=256, description="模板标题")
    category: str | None = Field(
        default=None, max_length=64, description="模板分类")
    description: str | None = Field(
        default=None, description="模板描述，Markdown格式")


class TemplateListItem(BaseModel):
    r"""
    模板列表项，不包含完整代码
    """

    id: int = Field(description="模板ID")
    title: str = Field(description="模板标题")
    category: str = Field(description="模板分类")
    description: str = Field(description="模板描述，Markdown格式")
    favorite_count: int = Field(description="收藏数")
    is_favorited: bool = Field(default=False, description="当前用户是否已收藏")
    created_at: datetime = Field(description="创建时间")


class TemplateDetail(BaseModel):
    r"""
    模板详情，包含所有语言的代码实现
    """

    id: int = Field(description="模板ID")
    title: str = Field(description="模板标题")
    category: str = Field(description="模板分类")
    description: str = Field(description="模板描述，Markdown格式")
    favorite_count: int = Field(description="收藏数")
    codes: list[TemplateCodeResponse] = Field(description="多语言代码实现列表")
    is_favorited: bool = Field(default=False, description="当前用户是否已收藏")
    created_at: datetime = Field(description="创建时间")
    updated_at: datetime = Field(description="更新时间")


class TemplateListResponse(BaseModel):
    r"""
    模板列表分页响应
    """

    items: list[TemplateListItem] = Field(description="模板列表")
    total: int = Field(description="总数")
    page: int = Field(description="当前页码")
    limit: int = Field(description="每页数量")

File: render/vue/src/types/template.ts
/**
 * 模板相关类型定义
 *
 * 与后端 /api/templates/* 接口一一对应，字段名使用 camelCase，
 * API 层负责 snake_case ↔ camelCase 的双向映射。
 *
 * @file src/types/template.ts
 * @author WaterRun
 * @date 2026-02-11
 */

import type { PaginationParams } from './api'

/**
 * 模板关联的单语言代码
 *
 * @interface
 */
export interface TemplateCode {
    /** 语言标识 */
    language: string
    /** 代码内容 */
    code: string
    /** 说明（可为 null） */
    explanation: string | null
}

/**
 * 模板完整信息（详情页）
 *
 * @interface
 */
export interface Template {
    /** 模板 ID */
    id: number
    /** 标题 */
    title: string
    /** 描述 */
    description: string
    /** 分类 */
    category: string
    /** 收藏数 */
    favoriteCount: number
    /** 当前用户是否已收藏 */
    isFavorited: boolean
    /** 各语言代码 */
    codes: TemplateCode[]
    /** 创建时间（ISO 8601） */
    createdAt: string
    /** 更新时间（ISO 8601） */
    updatedAt: string
}

/**
 * 模板摘要信息（列表页）
 *
 * @interface
 */
export interface TemplateListItem {
    /** 模板 ID */
    id: number
    /** 标题 */
    title: string
    /** 描述 */
    description: string
    /** 分类 */
    category: string
    /** 收藏数 */
    favoriteCount: number
    /** 当前用户是否已收藏 */
    isFavorited: boolean
    /** 创建时间（ISO 8601） */
    createdAt: string
}

/**
 * 模板搜索 / 筛选参数
 *
 * @interface
 */
export interface TemplateSearchParams extends PaginationParams {
    /** 搜索关键词 */
    keyword?: string
    /** 分类筛选 */
    category?: string
}